Adds DDO credentials.

calina-c opened the pull request:
Closes #296. Closes #298. Clean-up #277

calina-c commented:
Leaving as Draft and removing reviewers for now, to be combined with #298.

calina-c commented on ocean_lib/common/agreements/consumable.py:
```diff
@@ -0,0 +1,20 @@
+    CREDENTIAL_IN_DENY_LIST = 4
+
+
+class UnsupportedCredential(Exception):
```
remove. If it is not an address, do not throw the unsupported credential error

calina-c commented on ocean_lib/common/ddo/ddo.py:
```diff
@@ -354,3 +376,49 @@ def create_authentication_from_json(values):
+
+        return ConsumableCodes.OK
+
+    def is_consumable(self, credential=None, provider_uri=None):
```
flag for "with_connectivity" 🆗

DMats commented on ocean_lib/models/test/test_ddo.py:
```diff
@@ -37,6 +39,93 @@ def get_ddo_sample(datatoken_address):
+    assert ddo.get_addresses_of_type("allow") == ["0x123", "0x456a"]
+    assert ddo.get_addresses_of_type("deny") == ["0x2222", "0x333"]
+    assert (
+        ddo.get_address_allowed_code({"type": "address", "value": "0x111"})
```
Based on my recommendation to remove the `get_address_allowed_code` method, I recommend that you use `is_consumable` here like so:

```python
assert (
    ddo.is_consumable({"type": "address", "value": "0x111"}, with_connectivity_check=False) == ConsumableCodes.CREDENTIAL_NOT_IN_ALLOW_LIST
)
```

Extrapolate this recommendation to all other uses of `get_address_allowed_code`

calina-c replied:
Addressed that in separation of concerns. And, as you can see, tests are also easier to do when functions handle separate concerns. I am interested in testing ONLY address access here, not connectivity or disabling. Using your proposed version of `is_consumable`, I would have to bundle up all these tests together.

calina-c commented on ocean_lib/common/agreements/consumable.py:
```diff
@@ -0,0 +1,21 @@
+    CREDENTIAL_IN_DENY_LIST = 4
+
+
+class MalformedCredential(Exception):
```
That would only leave the ConsumableCodes declared here. And the two are tightly linked (2 of the codes are `CREDENTIAL`-related). I would keep them in the same file.

DMats replied:
Fair enough.

calina-c commented on ocean_lib/common/ddo/ddo.py:
```diff
@@ -354,3 +387,64 @@ def create_authentication_from_json(values):
         authentication = {"type": authentication_type, "publicKey": key_id}
         return authentication
+
+    def get_addresses_of_type(self, section_type="allow"):
```
I wouldn't go into generalising quite yet. That is why I only left some comments indicating that there can be extensions in the future. Why?
1. this PR already does a lot of things
2. we don't know which transformations apply to other credentials. E.g. address must be lowercased for any comparison, but maybe other credentials are case sensitive or need some symbol stripping etc. Which is why we can't generalise _now_.

Ref. your point (1), I agree the naming is bad now, but I'm not in love with `allow_or_deny` either, because it contains an `or`, and it also restricts. Will all future credentials only have `allow` and `deny`, or could there be other modifiers? I'll add some clarity here, but I need to give it some more thought.

DMats replied:
> we don't know which transformations apply to other credentials. E.g. address must be lowercased for any comparison, but maybe other credentials are case sensitive or need some symbol stripping etc. Which is why we can't generalise now.

I totally missed this in my initial review.  You're right. It's too early to generalize.

> Will all future credentials only have allow and deny, or could there be other modifiers? 

Standard ACL rules used in networking are either "permit" or "deny". Other modifiers are treated as separate parameters. That said, I like your chosen variable name "access_class" which resolves my concern about the variable name.

calina-c commented on ocean_lib/common/ddo/ddo.py:
```diff
@@ -354,3 +387,64 @@ def create_authentication_from_json(values):
+        if not self._status:
+            self._status = {}
+
+        self._status.pop("isOrderDisabled")
```
I went for implicit rather than explicit. If that key is not present, I think the structure of the DDO is clearer, otherwise a lot of flags will be set even if not needed. What do you think?

calina-c replied:
I went for implicit rather than explicit. If that key is not present, I think the structure of the DDO is clearer, otherwise a lot of flags will be set even if not needed. What do you think?

calina-c commented on ocean_lib/common/ddo/ddo.py:
```diff
@@ -354,3 +387,64 @@ def create_authentication_from_json(values):
+
+        self._status["isOrderDisabled"] = True
+
+    def get_address_allowed_code(self, credential=None):
```
I think it makes it much more easier to read, by separating the concerns. One function handles high-level access granting, while this one is a helper. No, it should not be called outside is_consumable, but we don't have protected/private in Python :(

DMats replied:
Fair enough. In that case, I recommend prepending an underscore like so:  `_get_allowed_code` to hint that this method is meant for internal use. I based this name on the latest changes which have already removed the "address" part of the method name

DMats replied:
I'm resolving this comment because this method is no longer internal because it's now located in `credentials.py` but it's called from `ddo.py`. Thus the prepended underscore no longer makes sense.

calina-c commented on ocean_lib/common/ddo/ddo.py:
```diff
@@ -354,3 +387,64 @@ def create_authentication_from_json(values):
+
+        return ConsumableCodes.OK
+
+    def requires_address_credential(self):
```
See my comment above regarding generalising.

calina-c commented on ocean_lib/common/ddo/ddo.py:
```diff
@@ -354,3 +387,64 @@ def create_authentication_from_json(values):
+        return allowed_addresses or denied_addresses
+
+    def is_consumable(
+        self, credential=None, with_connectivity_check=True, provider_uri=None
```
I don't want to do this. I started like this, but CodeClimate gave me really bad ratings on readability and I agree. See my comment above on separation of concerns.

DMats replied:
Fair enough.

calina-c commented on ocean_lib/common/utils/utilities.py:
```diff
@@ -73,3 +75,13 @@ def checksum(seed):
     return f"{datetime.utcnow().replace(microsecond=0).isoformat()}Z"
+
+
+def simplify_credential_to_address(credential):
```
This function does one thing: takes the "value" out of the credential object :)

DMats replied:
Based on your assertion that this method takes the "value" out of the credential object, I'd argue that it shouldn't exist because it should be as simple as `credential["value"]`

But if you insist on keeping this method, then I think it should be moved to `common/agreements/consumable.py` so that all the inputs and outputs of credential-related logic can be stored together. I don't think it belongs in `utilities.py` because the rest of the file focuses on id creation, converting web3 instances to blobs of bytes or hex, checksums, and timestamps.

Based on our discussion earlier today, `consumable.py` is separate from the credential-related logic so that users don't need to import the whole `AddressCredential` or `DDO` classes. I'm fine with this.

DMats replied:
I'll mark this resolved because I see `simplify_credential_to_address` was moved to `ocean_lib/common/ddo/credentials.py`

calina-c commented on ocean_lib/data_provider/data_service_provider.py:
```diff
@@ -312,56 +312,6 @@ def stop_compute_job(did, job_id, service_endpoint, consumer_address, signature)
         )
 
-    @staticmethod
-    def restart_compute_job(
```
The restart functionality was previously removed from higher-level components, and it was left stale here (not called anywhere).

calina-c replied:
Ref: https://github.com/oceanprotocol/ocean.py/commit/1dab70d164ca36a6cff284e8be82ae04344ad13f

DMats replied:
I added a note to the description about how this PR Cleans up #277

calina-c commented:
@DMats I refactored the whole shabang for a middle ground between your review and my... "reluctance" 😁 
I only left some convenience methods related to addresses in ddo.py, and moved most of the functionality to an AddressCredential class. When we decide to extend with other credentials, we can implement similar classes, and follow a factory-like pattern to select the Credential type.

Let me know, and let's touch base again when it's daytime in your timezone.

DMats commented on ocean_lib/common/ddo/credentials.py:
```diff
@@ -0,0 +1,106 @@
+
+        return ConsumableCodes.OK
+
+    def add_address_to_list_class(self, address, list_class="allow"):
```
Consider standardizing all uses of "list_class" to "access_class" because it's more descriptive.

DMats commented on ocean_lib/common/ddo/credentials.py:
```diff
@@ -0,0 +1,106 @@
+
+        return []
+
+    def requires_credential(self):
```
Add docstrings.

DMats commented on ocean_lib/common/ddo/credentials.py:
```diff
@@ -0,0 +1,106 @@
+    return credential["value"]
+
+
+def get_list_entry_with_type_address(types_list):
```
I think this should be a class method instead of a module function. And it should match the conventions set by `get_addresses_of_class`.  In fact, `get_addresses_of_class` should use this.

```python
    def get_address_entry_of_class(self, access_class="allow"):
        """Get a filtered list of addresses of a given access_class from credentials (use with allow/deny)."""
        entries = self.asset._credentials.get(access_class, [])
        return [entry for entry in entries if entry["type"] == "address"]

    def get_addresses_of_class(self, access_class="allow"):
        """Get an address entry of a given access class from credentials (use with allow/deny)."""
        address_entry = get_address_entry_of_class(access_class)
        return [val.lower() for val in address_entry["values"]]
```

DMats replied:
My apologies, I just realized my previous comment mixed up the docstrings.

DMats replied:
I'll just resolve this comment because I think it's become stale compared to the latest commits.

DMats commented on ocean_lib/common/ddo/credentials.py:
```diff
@@ -0,0 +1,106 @@
+    def __init__(self, asset):
+        self.asset = asset
+
+    def get_addresses_of_class(self, access_class="allow"):
```
Refactor based on the comment for `get_list_entry_with_type_address`

DMats replied:
Resolving because this comment has become stale. I'll mention it in my next review.

DMats commented on ocean_lib/common/ddo/credentials.py:
```diff
@@ -0,0 +1,106 @@
+            )
+            return
+
+        address_entry = address_entry[0]
```
Why is this line necessary?

Same question for the identical line in `remove_address_from_list_class`

calina-c replied:
I didn't pluralise corectly. I need to test that there is an address type before I modify the one addres type. It will be clearer after my next commit.

DMats replied:
Ohh I see now.  `get_address_entry_of_class` currently returns a list from which you extract the 0th entry.

I think `get_address_entry_of_class` should extract the 0th entry before returning because there should only ever be 1 address entry in a given access class list.

```python
    def get_address_entry_of_class(self, access_class="allow"):
        """Get the address entry of a given credentials access class list. access_class = "allow" or "deny"."""
        entries = self.asset._credentials.get(access_class, [])
        address_entries = [entry for entry in entries if entry["type"] == "address"]
        return address_entries[0] if address_entries else None
```

DMats commented on ocean_lib/common/ddo/credentials.py:
```diff
@@ -0,0 +1,106 @@
+
+def simplify_credential_to_address(credential):
+    if not credential:
+        return None
```
Shouldn't a MalformedCredential error be raised at this point as well?

calina-c replied:
no "Credential" means no dictionary as well, i.e. there is no address credential, nor other type of credential. In the future we might chain multiple credentials, so I check for the "requires address" separately.

DMats replied:
I don't understand your last comment. It sounds like you're talking about the `credentials` (plural) data structure described by https://github.com/oceanprotocol/multi-repo-issue/issues/98.

The `simplify_credential_to_address` function extracts the "value" from a `credential` (singular) which looks like `{"type": "something", "value": "something"}`.  If the caller passes anything that doesn't match that schema, this function should raise a `MalformedCredential` error.

calina-c replied:
The credential can be empty and the address can be None, it still needs checking. E.g. explicitly or implicitly passing empty. We still have to check if they are allowed. The case is covered by the same functions for allow and deny. 

Maybe I can't "see" these anymore, since I've spun them around too much. I'll look into it this morning. I'll submit answers for the other two comments.

calina-c commented on ocean_lib/common/ddo/credentials.py:
```diff
@@ -0,0 +1,96 @@
+
+
+class AddressCredential:
+    def __init__(self, asset):
```
I think it can be Asset too. I don't want to risk it for now 😟

DMats commented on ocean_lib/common/ddo/credentials.py:
```diff
@@ -0,0 +1,112 @@
+        lc_addresses.remove(address)
+        address_entry["values"] = lc_addresses
+
+    def get_address_entry_of_class(self, access_class="allow"):
```
I think `get_address_entry_of_class` should extract the 0th entry before returning because there should only ever be 1 address entry in a given access class list.

Also I just realized that I mixed up the docstrings between `get_address_entry_of_class` and `get_lc_addresses_from_entry`. So I decided to just re-write it to be a little more clear.

```python
    def get_address_entry_of_class(self, access_class="allow"):
        """Get address credentials entry of the specified access class. access_class = "allow" or "deny"."""
        entries = self.asset._credentials.get(access_class, [])
        address_entries = [entry for entry in entries if entry["type"] == "address"]
        return address_entries[0] if address_entries else None
```

DMats commented on ocean_lib/common/ddo/credentials.py:
```diff
@@ -0,0 +1,112 @@
+    def __init__(self, asset):
+        self.asset = asset
+
+    def get_addresses_of_class(self, access_class="allow"):
```
`get_addresses_of_class` and `get_lc_addressess_from_entry` return the same thing - the list of allowed or denied addresses (lower case). I think that only `get_addresses_of_class` should exist and it should look like this:

```python
    def get_addresses_of_class(self, access_class="allow"):
        """Get a filtered list of addresses from credentials (use with allow/deny)."""
        address_entry = self.get_address_entry_of_class(access_class)
        return [addr.lower() for addr in address_entry["values"]] if address_entry else []
```

alexcos20 reviewed (approved):
lgtm

calina-c commented:
@DMats merging in order to finish provider work. If anything is pending or reappears, I'll address it in a new PR.
